src/cleancall.c

```diff
@@ -10,23 +10,31 @@ static SEXP callbacks = NULL;
 static void push_callback(SEXP stack) {
   SEXP top = CDR(stack);
 
+  SEXP local_lgl = PROTECT(Rf_allocVector(LGLSXP, 1));
   SEXP fn_extptr = PROTECT(cleancall_MakeExternalPtrFn(NULL, R_NilValue, R_NilValue));
-  SEXP data_extptr = PROTECT(R_MakeExternalPtr(NULL, R_NilValue, R_NilValue));
+  SEXP data_extptr = PROTECT(R_MakeExternalPtr(NULL, local_lgl, R_NilValue));
   SEXP cb = Rf_cons(Rf_cons(fn_extptr, data_extptr), top);
 
   SETCDR(stack, cb);
 
-  UNPROTECT(2);
+  UNPROTECT(3);
 }
 
+struct data_wrapper {
+  SEXP (*fn)(void *data);
```

Comment:
General comment: `*` should stick to the type rather then the binding in our C style, right?

```suggestion
  SEXP (*fn)(void* data);
```